// azure-pipelines-tasks-codeanalysis-common/Common/FindbugsTool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FindbugsTool = void 0;
const BuildOutput_1 = require("./BuildOutput");
const BaseTool_1 = require("./BaseTool");
const path = require("path");
const fs = require("fs");
const xml2js = require("xml2js");
const tl = require("azure-pipelines-task-lib/task");
/**
 * An object that is able to configure the build to run FindBugs and identify and parse FindBugs reports
 *
 * @export
 * @class PmdReportParser
 * @implements {IAnalysisToolReportParser}
 */
class FindbugsTool extends BaseTool_1.BaseTool {
    constructor(buildOutput, boolInputName) {
        super('FindBugs', buildOutput, boolInputName);
    }
    /**
     * Configures the provided ToolRunner instance with arguments which will invoke the tool represented by this class.
     * @param toolRunner
     * @returns {ToolRunner} ToolRunner instance with arguments applied
     */
    configureBuild(toolRunner) {
        if (this.isEnabled()) {
            console.log(tl.loc('codeAnalysis_ToolIsEnabled'), this.toolName);
            switch (this.buildOutput.buildEngine) {
                case BuildOutput_1.BuildEngine.Maven:
                    toolRunner.arg(['findbugs:findbugs']);
                    break;
                case BuildOutput_1.BuildEngine.Gradle:
                    let initScriptPath = path.join(__dirname, '..', 'findbugs.gradle');
                    toolRunner.arg(['-I', initScriptPath]);
                    break;
                default:
                    break;
            }
        }
        return toolRunner;
    }
    /**
     * Implementers must specify where the XML reports are located
     */
    getBuildReportDir(output) {
        switch (this.buildOutput.buildEngine) {
            case BuildOutput_1.BuildEngine.Maven:
                return path.join(output.moduleRoot);
            case BuildOutput_1.BuildEngine.Gradle:
                return path.join(output.moduleRoot, 'reports', 'findbugs');
            default:
                tl.debug('No such build engine ' + this.buildOutput.buildEngine);
                throw new Error();
        }
    }
    /**
     * Report parser that extracts the number of affected files and the number of violations from a report
     *
     * @returns a tuple of [affected_file_count, violation_count]
     */
    parseXmlReport(xmlReport, moduleName) {
        let jsonCounts = [0, 0];
        let reportContent = fs.readFileSync(xmlReport, 'utf-8');
        xml2js.parseString(reportContent, (err, data) => {
            jsonCounts = FindbugsTool.parseJson(data);
            if (jsonCounts == null) {
                tl.debug(`[CA] Empty or unrecognized FindBugs XML report ${xmlReport}`);
                return null;
            }
            let violationCount = jsonCounts[1];
            // No files with violations, return now that it has been marked for upload
            if (violationCount === 0) {
                tl.debug(`[CA] A FindBugs report was found for module '${moduleName}' but it contains no violations`);
                return null;
            }
            tl.debug(`[CA] A FindBugs report was found for for module '${moduleName}' containing ${violationCount} issues - ${xmlReport}`);
        });
        return jsonCounts;
    }
    /**
     * Extracts the number of affected files and the number of violations from a report JSON object
     * @param data JSON object to parse
     * @returns a tuple of [affected_file_count, violation_count]
     */
    static parseJson(data) {
        // If the file is not XML, or is not from FindBugs, return immediately
        if (!data || !data.BugCollection ||
            !data.BugCollection.FindBugsSummary || !data.BugCollection.FindBugsSummary[0] ||
            !data.BugCollection.FindBugsSummary[0].PackageStats ||
            !data.BugCollection.FindBugsSummary[0].PackageStats[0].ClassStats) {
            return null;
        }
        let fileCount = 0;
        let violationCount = 0;
        // Extract violation and file count data from the sourceFile attribute of ClassStats
        let filesToViolations = new Map(); // Maps files -> number of violations
        data.BugCollection.FindBugsSummary[0].PackageStats[0].ClassStats.forEach((classStats) => {
            // The below line takes the sourceFile attribute of the classStats tag - it looks like this in the XML
            // <ClassStats class="main.java.TestClassWithErrors" sourceFile="TestClassWithErrors.java" ... />
            let sourceFile = classStats.$.sourceFile;
            let newBugCount = Number(classStats.$.bugs);
            if (newBugCount > 0) {
                // If there was not already an entry, start at 0
                if (!filesToViolations.has(sourceFile)) {
                    filesToViolations.set(sourceFile, 0);
                }
                // Increment bug count
                let oldBugCount = filesToViolations.get(sourceFile);
                filesToViolations.set(sourceFile, oldBugCount + newBugCount);
            }
        });
        // Sum violations across all files for violationCount
        for (let violations of filesToViolations.values()) {
            violationCount += violations;
        }
        // Number of <K,V> pairs in filesToViolations is fileCount
        fileCount = filesToViolations.size;
        return [violationCount, fileCount];
    }
}
exports.FindbugsTool = FindbugsTool;
